### PR TITLE
Suppress spam from calling canTransform

### LINKED ISSUE #501: tf2_ros::Buffer::canTransform sends many warnings if frame does not exist
When I call the tf2_ros::Buffer::canTransform function with a non-zero duration, I can't disable the warning even though I set the errstr argument.

In this line, tf2::BufferCore::canTransform is called without the errstr argument, and the warning is not disabled:
https://github.com/ros2/geometry2/blob/74b873a497f405f4296bb1c5de4023ff45eb9007/tf2_ros/src/buffer.cpp#L149

### PR DESCRIPTION
**Pull request by gonzodepedro**
Based on #502

Closes https://github.com/ros2/geometry2/issues/501

### CODE REVIEW

**Comment by jacobperron** [comment]
Do you mind also updating these other calls (from my other comment https://github.com/ros2/geometry2/pull/502#issuecomment-1104586927):

https://github.com/ros2/geometry2/blob/ef2dd147a25f14ed946c86f5a9fc03c8995d8fbe/tf2_ros/src/buffer.cpp#L94

https://github.com/ros2/geometry2/blob/ef2dd147a25f14ed946c86f5a9fc03c8995d8fbe/tf2_ros/src/buffer.cpp#L117

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16867/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11442/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17273/)

**Review by clalancette — COMMENTED** [review]
Not to derail this, but I will point out that suppressing these error messages unconditionally basically constitutes a large behavior change in tf2_ros from previous versions.  It also means that we won't get notified of real errors in the tf2 tree (like loops).

I'm not quite sure what we want to do here.  Maybe it would be a better idea to rate-limit this instead of suppressing it completely, as we are going to be losing a lot of debugging information this way.

**Comment by jacobperron** [comment]
The code is a bit tangled, but I think these are the two behavior changes caused by this PR:

1. Users calling `tf2_ros::Buffer::canTransform` with a `timeout` and `errstr` argument (that is not `nullptr`) now do not see a wall full of console warnings (3946b6e797a1fd04731411a1b0e4c2f7a39115d8). Instead, `errstr` is set to any warning.
2. Users calling `t2_ros::Buffer::lookupTransform` with a `timeout` argument now do not see a wall of console warnings (from the internal `canTransform()` invocation) (00fa84285691893c32786e1113b00aad81ebe2c6).

Correct me if I'm wrong, but I don't think (1) is disagreeable.

Regarding (2), the code is a bit hard to follow, but in both the previous and current behavior an exception is thrown from `lookupTransform` if there is an issue validating frames:

https://github.com/ros2/geometry2/blob/51a7f24191198eb9fc8124d36aba5bb2f7ad84f3/tf2/src/buffer_core.cpp#L124-L145

called from lookupTransform's implementation:

https://github.com/ros2/geometry2/blob/51a7f24191198eb9fc8124d36aba5bb2f7ad84f3/tf2/src/buffer_core.cpp#L654-L655

The only difference this PR introduces is that we don't get the (more or less) same message repeated to the console before the exception ultimately occurs. AFAICT, we're only calling `canTransform` inside `lookupTransform` as a lazy way to implement the timeout. Both `canTransform` and `lookupTransform` do frame ID validation, but the former prints warnings to console and the latter throws an exception.

I agree that the underlying issue of  `tf2::canTransform` printing to console at an unholy rate should probably be fixed.
However, IMO, it doesn't make sense to print warnings to console while we're waiting for a transform to become available with `lookupTransform`, especially since we're already raising an exception with the same message if there's still a problem when the timeout is reached.

**Comment by jacobperron** [comment]
If I'm not mistaken, this change doesn't hide errors in the TF tree since `lookupTransform` still throws an exception with the same error message that was previously printed to console.

**Comment by jacobperron** [comment]
@clalancette friendly ping :slightly_smiling_face:

**Comment by clalancette** [comment]
> Correct me if I'm wrong, but I don't think (1) is disagreeable.

Agreed, this is desirable.  The part of the patch that changes `canTransform` looks good.

> Regarding (2), the code is a bit hard to follow, but in both the previous and current behavior an exception is thrown from `lookupTransform` if there is an issue validating frames:

And now looking at it in more detail, I agree with you.  `canTransform` is basically being reused as a way to "wait" for the transform to come in, and as such should not print any warnings.  Once that expires and we get to the `lookupTransform` API (which actually is inherited from the underlying `tf2::BufferCore` class), then if that transform still doesn't exist we'll throw an exception.  That seems like sufficient warning to the user.

So with all of that said, I'm going to approve this.  Thanks for the detailed explanation on what is going on here.

**Review by clalancette — APPROVED** [review]

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17160/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11704/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17581/) (unrelated cmake warning from `osrf_testing_tools_cpp`)

**Comment by vincentrou** [comment]
@jacobperron 
Thanks for merging this PR.
Is it possible to backport this to humble ?
Since it is the last LTS and there is a lot of related issues : https://github.com/ros2/geometry2/issues/358

**Comment by leander2189** [comment]
Was this feature ever ported to humble? I am getting lots of messages in my tests, they make debugging a lot more difficult

**Comment by clalancette** [comment]
@mergifyio backport humble
